Simplify getting Encoding in `TranscriptionOption.FlushContentToDisk`

Gimly opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

Simplify getting encoding of existing file in the `FlushContentToDisk` method of the `TranscriptionOption` class called when the transcription file is flushed to disk.

## PR Context

The `FlushContentToDisk` method was calling a complex `GetEncoding` utils that was trying to read the file encoding "by hand". Through a discussion with @iSazonov in #13677 and #13899 it was decided that this could be simplified by using a StreamReader and taking advantage of the `CurrentEncoding` property.

This allows to reduce code complexity and should also improve compatibility with different encoding. This PR should have no functional impact except improve recognition of encoding of existing files in some cases not covered by the existing code.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

Gimly committed:
Remove Encoding property in TranscriptionOption

Gimly committed:
Simplify Utils.GetEncoding to use StreamReader

Utils.GetEncoding was using a complicated implementation that could be easily replaced
by StreamReader.GetCurrentEncoding.

Used that implementation in the FlushContentToDisk method

Gimly committed:
Catch a more specific exception

Gimly commented:
@iSazonov I think that's what you had in mind. I didn't remove the GetEncoding method from Utils because I thought it was still cleaner to have that code outside of the FlushContentToDisk method, but it's greatly simplified. 

I've removed the Encoding property and simply call the GetEncoding method at the start of the FlushContentToDisk, I don't think that should have an impact on performances and I think it's a bit cleaner than opening a file in the setter of a property.

I decided to open a separate PR from the #13899 because I thought this, while related, is maybe a bit more open for discussion, hope that's OK for you.

iSazonov commented on src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs:
```diff
@@ -1120,13 +1099,14 @@ internal void FlushContentToDisk()
                 {
                     if (_contentWriter == null)
                     {
+                        var encoding = Utils.GetEncoding(this.Path);
                         try
                         {
```
If we remove GetEncoding() we should put StreamReader to try block.

Gimly replied:
Done as proposed

iSazonov commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -1399,66 +1399,17 @@ internal static Encoding GetEncoding(string path)
             }
             catch (IOException)
             {
                 return ClrFacade.GetDefaultEncoding();
```
We could return utf8NoBom - we discussed about the default previously.

Gimly replied:
In the catch you mean? I've seen that `ClrFacade.GetDefaultEncoding()` defaults also to utf8NoBom and it's used in a few other places. So I thought it was a bit cleaner to use this method instead of the static. Or should we replace all the calls to `GetDefaultEncoding` by the static in this class?

Gimly replied:
Done in the update, used utf8NoBom and removed the use of the GetDefaultEncoding method

Gimly commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -1399,66 +1399,17 @@ internal static Encoding GetEncoding(string path)
             try
             {
-                using (FileStream stream = System.IO.File.OpenRead(path))
+                using (StreamReader sr = new StreamReader(path, true))
```
I'm lost now, I thought we wanted to use the `StreamReader` to get the current file encoding, if the file already existed. Then keep that encoding. 

If I use the constructor with the Encoding, it will open the file with utf8NoBOM regardless of what the file's encoding is really. The sr.Encoding will always return utf8NoBOM, which means that we don't actually need to use the StreamReader in the first place.

I'm surely missing something.

iSazonov replied:
@Gimly 

> If I use the constructor with the Encoding, it will open the file with utf8NoBOM regardless of what the file's encoding is really

If we set the detection flag StreamReader will always call [DetectEncoding()](https://source.dot.net/#System.Private.CoreLib/StreamReader.cs,627) at first read operation. If it can not detect the file encoding it will use default. So we should set the default.

Gimly replied:
You're right sorry, I misread StreamReader's documentation. I've pushed an update with the default

Gimly committed:
Further cleanup and fix PR comment

Remove the GetEncoding from Utils and two related static readonly fields.

Add the default to utf8NoBom in the stream reader
Put the stream reader in the try catch.

iSazonov commented on src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs:
```diff
@@ -1122,11 +1101,18 @@ internal void FlushContentToDisk()
                             _contentWriter = new StreamWriter(
                                 new FileStream(this.Path, FileMode.OpenOrCreate, FileAccess.ReadWrite, FileShare.Read),
-                                this.Encoding);
+                                currentEncoding ?? Utils.utf8NoBom);
```
Please initialize `currentEncoding`  in line 1104 and remove this `??`.

Gimly replied:
Yes, much better, done

iSazonov commented on src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs:
```diff
@@ -1122,11 +1101,18 @@ internal void FlushContentToDisk()
                     {
                         try
                         {
+                            Encoding currentEncoding = null;
```
```suggestion
                            Encoding currentEncoding = Utils.utf8NoBom;
```

Gimly replied:
done

iSazonov commented on src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs:
```diff
@@ -1122,11 +1101,18 @@ internal void FlushContentToDisk()
+
+                            using (StreamReader reader = new StreamReader(this.Path, Utils.utf8NoBom, true))
+                            {
+                                currentEncoding = reader.CurrentEncoding;
```
You lost a formal read operation to force encoding detection.

Gimly replied:
done

iSazonov commented on src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs:
```diff
@@ -1122,11 +1101,18 @@ internal void FlushContentToDisk()
                         {
+                            Encoding currentEncoding = null;
+
+                            using (StreamReader reader = new StreamReader(this.Path, Utils.utf8NoBom, true))
```
Please use named arguments for consts.
```suggestion
                            using (StreamReader reader = new StreamReader(this.Path, Utils.utf8NoBom, detectEncodingFromByteOrderMarks: true))
```

Gimly replied:
done

Gimly committed:
Implement PR suggestions

Gimly committed:
There is no need to initialize the variable

iSazonov commented on src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs:
```diff
@@ -1122,11 +1101,19 @@ internal void FlushContentToDisk()
                     {
                         try
                         {
+                            Encoding currentEncoding;
```
Since we don't initialize the variable we can move the definition to line 1109

Gimly replied:
Because of the using we can't, and I can't put the rest of the code inside of the using otherwise the StreamWriter will not have access to the file because it will be in use.

Gimly commented:
@iSazonov Added three unit tests, for the version without the Append I tested with an existing file with another encoding, I think that was the main issue.

I see that the mac os CI have failed, not sure what's happening there. An Invoke-WebRequest that has failed.

rjmholt commented on src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs:
```diff
@@ -1122,11 +1101,19 @@ internal void FlushContentToDisk()
                     {
                         try
                         {
+                            Encoding currentEncoding;
+
+                            using (StreamReader reader = new StreamReader(this.Path, Utils.utf8NoBom, detectEncodingFromByteOrderMarks: true))
+                            {
+                                _ = reader.Read();
+                                currentEncoding = reader.CurrentEncoding;
+                            }
+
```
It seems like this is actually interspersing two pieces of logic:

- Determining the encoding we want to use
- Creating the file stream writer

If that's the case, it might be nicer to factor the first into its own method, similar to the pre-existing code:

```csharp
private Encoding GetPathEncoding()
{
    using (var reader = new StreamReader(Path, Utils.utf8NoBom, detectEncodingFromByteOrderMarks: true))
    {
        _ = reader.Read();
       return reader.CurrentEncoding;
    }
}

internal void FlushContentToDisk()
{
    lock (OutputBeingLogged)
    {
        try
        {
            // Does this dispose check need to be within the lock at all?
            if (disposed)
            {
                return;
            }

            if (_contentWriter != null)
            {
                return;
            }

            try
            {
                _contentWriter = new StreamWriter(
                    new FileStream(Path, FileMode.OpenOrCreate, FileAccess.ReadWrite),
                    GetPathEncoding());

                // Worth adding a comment here to explain what's happening
                _contentWriter.BaseStream.Seek(0, SeekOrigin.End);
            }
            catch (IOException)
            {
                _contentWriter = new StreamWriter(
                    new FileStream(Path, FileMode.Append, FileAccess.Write, FileShare.Read),
                    Utils.utf8NoBom);
            }
        }
        finally
        {
            OutputBeingLogger.Clear();
        }
    }
}
```

Gimly replied:
That's what I had initially, with the `GetPathEncoding` defined in the `Utils` class, but I think @iSazonov didn't like the method there. 

In the end, it will be used only in this `FlushContentToDisk`, maybe it should be a local function? This way it is factored, but still available only in this method?

Otherwise I can put it back into Utils, in the end it's still valid, even more so if we just let the `IOException` bubble if there's an issue reading the file.

iSazonov replied:
If a short method is used once I don't see a point to have it (especially since it should be in
try-catch block). We could add a comment before the code block.

Gimly replied:
@rjmholt what do you think?

rjmholt replied:
Well I don't like making you go around in circles. So I'm happy to leave the code as it is.

@iSazonov the current PowerShell code is full of long tracts of statements, and breaking things into methods is an easy way to get readability and also improves reusability later on. Honestly I just find it to simplify the process for writing and reviewing code. That's where my preference comes from. And given the complexity of some parts of our codebase, I don't see much argument against it.

> especially since it should be in try-catch block

This is the kind of clarity that putting something into a method provides. As is, it's not clear what the full intent and purpose of the try/catch is (why do we fall back to a totally different setting in the catch block?). Once you separate the logic into two pieces and must give them a human-readable name, it's easier to assess what each operation's error semantics should be.

Gimly replied:
I don't mind changing things, in the end the goal was to make the code more readable / simple, and I tend to agree that it would probably be clearer in a method, which makes it clear why we're creating this StreamReader here.

Since what seems to bother @iSazonov is the fact that the method would be used only in this method, I think it's a good candidate for a local function, it groups the logic while keeping it "inside" the method and usable only there. Do you think this would be an acceptable proposal?

iSazonov replied:
@rjmholt Thanks for clarify!
I see your point and agree that a method with an informative name adds readability. But in this case it seemed to me unnecessary. We have a request to create a cmdlet like `Get-FileEncoding` but I don't think this piece of code is worth worrying about reuse.
As for this try-catch block, I was afraid that we could change something in the behavior and get regression now or in the future.

Gimly replied:
So, as discussed I did a test with a local function. I think it's a good middle ground between your two opinions, we have the code factored and with an understandable name and at the same time it's available only in the `FlushContentToDisk` method.
Waiting on your remarks.

rjmholt replied:
I think a local function sounds ideal 🙂

Gimly committed:
Factor get path encoding

Use a local function to factor the path encoding code.

iSazonov commented on src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs:
```diff
@@ -1149,6 +1130,15 @@ internal void FlushContentToDisk()
                 OutputBeingLogged.Clear();
             }
+
+            Encoding getPathEncoding()
```
Please move the local function to start of the method.
Also please use Pascal name case - local function is de-facto a method and could be move in another scope.

Gimly replied:
Done and done. Interesting that you prefer the local function at the start. I haven't seen any recommendation for local functions yet. Do you happen to have coding conventions available for them?

In my team we prefer to put it at the bottom, this way you have "pseudo-code" at the beginning with only the calls and you can just scroll down to see the implementation details.

Gimly committed:
Moved local function to start of method and PascalCase

iSazonov commented on src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs:
```diff
@@ -1101,19 +1086,22 @@ internal string Path
         /// </summary>
         internal void FlushContentToDisk()
         {
+            Encoding GetPathEncoding()
+            {
+                using (StreamReader reader = new StreamReader(this.Path, Utils.utf8NoBom, detectEncodingFromByteOrderMarks: true))
+                {
+                    _ = reader.Read();
+                    return reader.CurrentEncoding;
+                }
+            }
```
Recently we activated a Roslyn analyzer which suggest to make this static if possible. Other suggest simplify `using`.
```suggestion
            static Encoding GetPathEncoding(string path)
            {
                using StreamReader reader = new StreamReader(path, Utils.utf8NoBom, detectEncodingFromByteOrderMarks: true);
                _ = reader.Read();
                return reader.CurrentEncoding;
            }
```

Gimly committed:
Make GetPathEncoding static and use inline using

Co-authored-by: Ilya <darpa@yandex.ru>

iSazonov commented on src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs:
```diff
@@ -1122,11 +1108,13 @@ internal void FlushContentToDisk()
                     {
                         try
                         {
+                            var currentEncoding = GetPathEncoding();
```
To fix compiling:
```suggestion
                            var currentEncoding = GetPathEncoding(this.Path);
```

Gimly committed:
Add back missing parameter to GetPathEncoding Path

Co-authored-by: Ilya <darpa@yandex.ru>

iSazonov commented:
@Gimly Thanks for your contribution!
